### ISSUE TITLE
Attention model

### ISSUE THREAD

**Issue by fuleying** [fuleying/fuleying.github.io#6]
### show, attend, and tell
![image](https://user-images.githubusercontent.com/6255387/46084583-93b3db00-c1d6-11e8-9534-3cb53602a77d.png)
### Spatial Attention Model
![image](https://user-images.githubusercontent.com/6255387/46084653-bf36c580-c1d6-11e8-829d-96f7ed31f110.png)
### [Adaptive Attention](https://arxiv.org/pdf/1612.01887.pdf)
![image](https://user-images.githubusercontent.com/6255387/46084700-d83f7680-c1d6-11e8-95dd-55ccd20d5643.png)

**Comment by fuleying** [comment]
![image](https://user-images.githubusercontent.com/6255387/46118427-0e154700-c239-11e8-8ebf-2dd576091bfd.png)

**Comment by fuleying** [comment]
### hard attention
![image](https://user-images.githubusercontent.com/6255387/46120914-0018f380-c244-11e8-9888-b0de56be437c.png)

**Comment by fuleying** [comment]
### [Attention in Translation](https://medium.com/syncedreview/a-brief-overview-of-attention-mechanism-13c578ba9129)
![image](https://user-images.githubusercontent.com/6255387/46145807-d687b880-c293-11e8-94a6-c283e683c30f.png)
- Blue represents encoder and red represents decoder.
- The context vector takes `all cells’ outputs as input` to compute the probability distribution of source language words for `each single word decoder wants to generate`. 
- By utilizing this mechanism, it is possible for decoder to `capture somewhat global information` rather than solely to infer based on one hidden state.

**Comment by fuleying** [comment]
### build context vector
![image](https://user-images.githubusercontent.com/6255387/46145992-3d0cd680-c294-11e8-8599-98c939c8e985.png)
- For a fixed target word, first, we `loop over all encoders’ states` to **compare target and source states to generate scores** for each state in decoders. 
- Then we could `use softmax to normalize all scores`, which generates the probability distribution conditioned on target states. 
- At last, the weights are introduced to make `context vector` easy to train. 

1. **During decoding, context vectors are computed for every output word.** So we will have a `2D matrix` whose size is **# of target words multiplied by # of source words**. Equation (1) demonstrates how to compute a single value `given one target word and a set of source word.`
2. Once context vector is computed, `attention vector could be computed by context vector, target word, and attention function f.`
3. We need attention mechanism to be **trainable**. According to equation (4), both styles offer the trainable weights ([W in Luong’s](https://arxiv.org/abs/1508.04025), [W1 and W2 in Bahdanau’s](https://arxiv.org/abs/1409.0473)). Thus, different styles may result in different performance.

**Comment by fuleying** [comment]
### GAT Attention
- N个结点(feature_size = F)之间的attention：`N×N`
- 先通过变化变成`[N, N, Feature_size]`, 然后再通过MLP降维成`[N, N, 1]`，squeeze成`[N, N]`
- 然后就可以通过softmax得到N个结点之间的attention `[N, N]`
![image](https://user-images.githubusercontent.com/6255387/76766873-a8719d80-67d3-11ea-9dce-50ea15a5651f.png)
https://github.com/Diego999/pyGAT/blob/e6a8fa50a01202588c2d3980356d7df216f69957/layers.py#L7-L46

```python
def forward(self, input, adj):
    # input: [N, F]
    # adj: [N, N]
    # h: [N, F']
    h = torch.mm(input, self.W)
    N = h.size()[0]

    a_input = torch.cat(
        # [N, NF'] -> [NN, F']    [NN, F']
        [h.repeat(1, N).view(N * N, -1), h.repeat(N, 1)], dim=1
    ).view(N, -1, 2 * self.out_features) # [N, N, 2F']
    self.a = nn.Parameter(torch.zeros(size=(2 * self.out_features, 1)))
    # [N, N, 2F'] -> [N, N, 1] -> [N, N]
    e = self.leakyrelu(torch.matmul(a_input, self.a).squeeze(2))

    zero_vec = -9e15 * torch.ones_like(e)
    attention = torch.where(adj > 0, e, zero_vec)
    # [N, N]
    attention = F.softmax(attention, dim=1)
    attention = F.dropout(attention, self.dropout, training=self.training)
    # [N, N] * [N, F'] -> [N, F']
    h_prime = torch.matmul(attention, h)

    if self.concat:
        return F.elu(h_prime)
    else:
        return h_prime
```

**Comment by fuleying** [comment]
## Self-attention

```python
class Attention(nn.Module):
    def __init__(self, q_dim, k_dim, v_dim, output_size, **kwargs):
        super(Attention, self).__init__()
        self.q_dim = q_dim
        self.k_dim = k_dim
        self.v_dim = v_dim
        self.output_size = output_size
        self.use_residual = kwargs.get("use_residual", True)
        self.qk_hidden = kwargs.get("qk_hidden", 16)

        self.Wq = nn.Linear(self.q_dim, self.qk_hidden)
        nn.init.xavier_uniform_(self.Wq.weight, gain=1.414)

        self.Wr = nn.Sequential(nn.Linear(self.q_dim, self.output_size), nn.ReLU())
        nn.init.xavier_uniform_(self.Wr[0].weight, gain=1.414)

        self.Wk = nn.Linear(self.k_dim, self.qk_hidden)
        nn.init.xavier_uniform_(self.Wk.weight, gain=1.414)

        self.Wv = nn.Sequential(nn.Linear(self.v_dim, self.output_size), nn.ReLU())
        nn.init.xavier_uniform_(self.Wv[0].weight, gain=1.414)

        self.temp = self.qk_hidden ** 0.5

    def forward(self, q, k, v, masks=None):
        """
        :param q, k, v: [B, N, T, q_dim]
        :return: [B, N, T, output_size]
        """

        residual = self.Wr(q)
        q = self.Wq(q)  # [B, N, T, qk_hidden]
        k = self.Wk(k)  # [B, N, T, qk_hidden]
        v = self.Wv(v)  # [B, N, T, output_size]

        k = torch.transpose(k, -2, -1)
        # [B, N, T, T]
        attention = torch.matmul(q / self.temp, k)
        if masks is not None:
            attention = attention.masked_fill(masks <= 0, -1e9)
        attention = torch.softmax(attention, dim=-1)
        #         import random
        #         if random.random()<0.01:
        #             torch.save(attention,'attention.t')

        if self.use_residual:
            q = torch.matmul(attention, v) + residual
        else:
            # [B, N, T, output_size]
            q = torch.matmul(attention, v)

        return q

```
